openvdb/openvdb/points/PointStatistics.h

```diff
@@ -0,0 +1,609 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+//
+/// @author Nick Avramoussis
+///
+/// @file PointStatistics.h
+///
+/// @brief Functions to perform multi threaded reductions and analysis of
+///   arbitrary point attribute types. Each function imposes various
+///   requirements on the point ValueType (such as expected operators) and
+///   supports arbitrary point filters.
+///
+
+#ifndef OPENVEB_POINTS_STATISTICS_HAS_BEEN_INCLUDED
+#define OPENVEB_POINTS_STATISTICS_HAS_BEEN_INCLUDED
+
+#include "PointDataGrid.h"
+
+#include "../openvdb.h"
+#include "../Types.h"
+#include "../math/Math.h"
+#include "../tree/LeafManager.h"
+
+#include <tbb/parallel_reduce.h>
+#include <tbb/parallel_for.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+/// @brief Evaluates the minimum and maximum active values of a point attribute.
+/// @details  This function sets the arguments min and max to the minimum and
+///   and maximum values of a given point attributes. The ValueType of the
+///   attribute must be copy constructible and support less than and greater
+///   than operators. The ValueType does not need to support zero initialization
+///   or define its own numerical limits. This method will throw only if the
+///   templated ValueType does not match the given attribute. This method will
+///   return true if min and max have been set, false otherwise. The function is
+///   deterministic.
+/// @note  The value type of the min/max calculations must match the value type
+///   of the attribute. For vectors and matrices, this results in per component
+///   comparisons. See evalExtents for magnitudes or more custom control.
+/// @param points     the point tree
+/// @param attribute  the attribute to reduce
+/// @param min        the computed min value
+/// @param max        the computed max value
+/// @param filter     a filter to apply to points
+/// @param minTree    if provided, builds a tiled tree of localised min results
+/// @param maxTree    if provided, builds a tiled tree of localised max results
+/// @return true if min and max have been set, false otherwise. Can be false if
+///   no points were processed or if the tree was empty.
+/// @warning if "P" is provided as the attributes, it is evaluated in voxel space
+template <typename ValueT,
+    typename CodecT = UnknownCodec,
+    typename FilterT = NullFilter,
+    typename PointDataTreeT = points::PointDataTree>
+inline bool evalMinMax(const PointDataTreeT& points,
+        const std::string& attribute,
+        ValueT& min,
+        ValueT& max,
+        const FilterT& filter = NullFilter(),
+        typename PointDataTreeT::template ValueConverter<ValueT>::Type* minTree = nullptr,
+        typename PointDataTreeT::template ValueConverter<ValueT>::Type* maxTree = nullptr);
```

Comment:
Can you elaborate on why you think this is "clunky"? These are optional arguments e.g:
```
evalMinMax<int32_t>(tree, min, max, NullFilter());
...
Int32Tree minTree;
evalMinMax<int32_t>(tree, min, max, NullFilter(), &minTree);
```
I don't want to pollute the public API where possible.

> You could compute the former directly from the later pretty inexpensively?

Yes, but if you don't want these trees you shouldn't have to pay the penalty (more memory and slightly slower) to create them. In other words, they should not be constructed if not provided and we can utilise a full reduction rather than an extra post process.

Reply:
Added basic function singatures!